### PR TITLE
Fixed an bug which would not unpersist NormalizationContexts created for random effect problems

### PR DESCRIPTION
**Pull request by ashelkovnykov**
- NormalizationContexts are now created by the GameEstimator and re-used between optimization problems, then unpersisted at the end of GameEstimator execution
- Moved the Coordinate object creation logic to a new CoordinateFactory class, so that the GameEstimator.train function would be less bloated
- Various minor modifications to existing classes for the above changes
- A few new classes for the above changes
- New unit tests for the aforementioned new classes

### CODE REVIEW

File: photon-api/src/main/scala/com/linkedin/photon/ml/function/glm/GLMLossFunction.scala [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+   */
+  def buildFactory(
+      sc: SparkContext,
+      lossFunction: PointwiseLossFunction,
+      treeAggregateDepth: Int): (CoordinateOptimizationConfiguration) => ObjectiveFunction =
+
+    { config: CoordinateOptimizationConfiguration =>
+      config match {
+        case fEOptConfig: FixedEffectOptimizationConfiguration =>
+          DistributedGLMLossFunction(sc, fEOptConfig, lossFunction, treeAggregateDepth)
```
Comment (alice2008) [inline photon-api/src/main/scala/com/linkedin/photon/ml/function/glm/GLMLossFunction.scala]: my local build failed due to `too many arguments for method apply`

Comment (ashelkovnykov) [reply photon-api/src/main/scala/com/linkedin/photon/ml/function/glm/GLMLossFunction.scala]: I made an error while rebasing - try now

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala [hunk 2]
```diff
@@ -614,6 +647,37 @@ class GameEstimator(val sc: SparkContext, implicit val logger: Logger) extends P
       }
   }
 
+  protected def prepareNormalizationContextWrappers(
```
Comment (alice2008) [inline photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: Why there are many empty lines in this function?

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateFactory.scala [hunk 3]
```diff
@@ -0,0 +1,112 @@
+/*
```
Comment (alice2008) [inline photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateFactory.scala]: Missing unit tests for `CoordinateFactory`.

File: photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala (hunk 2 above)
Comment (ashelkovnykov) [reply photon-api/src/main/scala/com/linkedin/photon/ml/estimators/GameEstimator.scala]: This one is just an unfortunate function - regardless of what style convention is used, it looks like a mess

File: photon-api/src/integTest/scala/com/linkedin/photon/ml/function/DistributedObjectiveFunctionIntegTest.scala [hunk 4]
```diff
@@ -72,10 +72,10 @@ class DistributedObjectiveFunctionIntegTest extends SparkTestUtils {
       case TaskType.LOGISTIC_REGRESSION =>
```
[the remark this answers is not in the document]
Comment (ashelkovnykov) [reply photon-api/src/integTest/scala/com/linkedin/photon/ml/function/DistributedObjectiveFunctionIntegTest.scala]: Similar tests exist for the "Single Node" functions, in the `SingleNodeObjectiveFunctionTest`

File: photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateFactory.scala (hunk 3 above)
Comment (ashelkovnykov) [reply photon-api/src/main/scala/com/linkedin/photon/ml/algorithm/CoordinateFactory.scala]: Added

**Pull request merged**